In [1]:
import numpy as np
import math 
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pymysql                        # for getting data from a SQL database
from sqlalchemy import create_engine  # for establishing the connection and authentication

from getpass import getpass           # To get the password without showing the input

### Project topic: Approach to disprove Random-Walk theory with foreign exchange market data

#### Definition Random-Walk-Theory:
Says basically that it is impossible to predict the further progress and price of market data and therefore profits with specific trading strategies are not possible

#### Trading strategy:

- If low lower than low of 5 previous candles + candle is green (open < close), next day go long at opening
- If high higher than high of 5 previous candles + candle is red (open > close), next day go short at opening

--> reversed candle strategy

Rules:
- Positions will be opened always at open and closed at close of the day (no positions over night)

#### Setting up connection to SQL trading database

In [2]:
#Storing my SQL password in variable
password = getpass()

········


In [3]:
#Build the connection string to trading database and the engine
connection_string = 'mysql+pymysql://root:'+password+'@localhost/trading'
engine = create_engine(connection_string)

### Retrieving required data from trading database:
USDJPY daily will again be used as first trading pair

In [4]:
USDJPY = pd.read_sql_query('SELECT * FROM trading.usdjpy_day_new', engine)
USDJPY

date     open     high      low    close
0    2005-01-03  20548.0  20690.0  20464.0  20564.0
1    2005-01-04  20564.0  20958.0  20486.0  20888.0
2    2005-01-05  20888.0  21004.0  20752.0  20820.0
3    2005-01-06  20820.0  21036.0  20776.0  20984.0
4    2005-01-07  20984.0  21024.0  20772.0  20964.0
...         ...      ...      ...      ...      ...
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0

[5081 rows x 5 columns]

In [5]:
def calcOpenCloseDiff(df):
    df['diff_open_close'] = df.apply(lambda x: x['close'] - x['open'], axis=1)
    return df

In [6]:
USDJPY = calcOpenCloseDiff(USDJPY)
USDJPY

date     open     high      low    close  diff_open_close
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0
...         ...      ...      ...      ...      ...              ...
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0

[5081 rows x 6 columns]

#### Create function:
- compare low of current candle with lows of 5 previous candles
- if current candle low lower and open < close (green candle), insert in new column value '-1' --> signal for next day to go long
- if current candle high higher and open > close (red candle), insert in new column value '1' --> signal for next day to go short
- else: insert 0 --> no trade on next day

In [7]:
def classLongShort(df):
    list_class = []
    for index, row in df.iterrows():
        if (row['low'] < df['low'].shift(periods=1)[index]) & (row['low'] < df['low'].shift(periods=2)[index]) & (row['low'] < df['low'].shift(periods=3)[index]) & (row['low'] < df['low'].shift(periods=4)[index]) & (row['low'] < df['low'].shift(periods=5)[index]) & (row['diff_open_close'] > 0):
                list_class.append(-1)
        elif (row['high'] > df['high'].shift(periods=1)[index]) & (row['high'] > df['high'].shift(periods=2)[index]) & (row['high'] > df['high'].shift(periods=3)[index]) & (row['high'] > df['high'].shift(periods=4)[index]) & (row['high'] > df['high'].shift(periods=5)[index]) & (row['diff_open_close'] < 0):
                list_class.append(1)
        else:
            list_class.append(0)
    df['classification'] = list_class
    return df

In [8]:
%%time
classLongShort(USDJPY)
USDJPY

Wall time: 4.34 s


date     open     high      low    close  diff_open_close  \
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0   
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0   
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0   
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0   
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0   
...         ...      ...      ...      ...      ...              ...   
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0   
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0   
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0   
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0   
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0   

      classification  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
5076               1  
5077               0  
5078               0  
5079               0  
5080               0  

[5081 rows x 7 columns]

In [9]:
USDJPY['classification'].value_counts()

 0    4494
 1     325
-1     262
Name: classification, dtype: int64

#### Function to create column for order type for each day

In [10]:
def createOrderType(df):
    list_class = list(df['classification'])
    list_orderType = []
    for value in range(0, len(list_class)):
        if list_class[value-1] == -1:
            list_orderType.append('long')
        elif list_class[value-1] == 1:
            list_orderType.append('short')
        else:
            list_orderType.append('none')
    df['order_type'] = list_orderType
    return df

In [11]:
createOrderType(USDJPY)

date     open     high      low    close  diff_open_close  \
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0   
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0   
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0   
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0   
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0   
...         ...      ...      ...      ...      ...              ...   
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0   
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0   
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0   
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0   
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0   

      classification order_type  
0                  0       none  
1                  0       none  
2                  0       none  
3                  0       none  
4                  0       none  
...              ...        ...  
5076               1       none  
5077               0      short  
5078               0       none  
5079               0       none  
5080               0       none  

[5081 rows x 8 columns]

#### Create column with actual profit per day comparing order_type with diff_open_close

In [12]:
def createProfit(df):
    list_profit = []
    for index, row in df.iterrows():
        if row['order_type'] == 'long':
            list_profit.append(row['diff_open_close'])
        elif row['order_type'] == 'short':
            profit = (row['diff_open_close'])*(-1.0)
            list_profit.append(profit)
        else:
            list_profit.append(0)
    df['profit'] = list_profit
    return df

In [13]:
createProfit(USDJPY)

date     open     high      low    close  diff_open_close  \
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0   
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0   
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0   
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0   
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0   
...         ...      ...      ...      ...      ...              ...   
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0   
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0   
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0   
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0   
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0   

      classification order_type  profit  
0                  0       none     0.0  
1                  0       none     0.0  
2                  0       none     0.0  
3                  0       none     0.0  
4                  0       none     0.0  
...              ...        ...     ...  
5076               1       none     0.0  
5077               0      short    -1.0  
5078               0       none     0.0  
5079               0       none     0.0  
5080               0       none     0.0  

[5081 rows x 9 columns]

#### Calculate average loss, average win, number of trades and sum of profit

In [14]:
def loss_win_trades_profit(df):
    sum_of_profit = df['profit'].sum()
    number_of_trades = df['order_type'][(df['order_type'] == 'long') | (df['order_type'] == 'short')].count()
    average_loss = df['profit'][df['profit'] < 0].mean()
    average_win = df['profit'][df['profit'] > 0].mean()
    
    return sum_of_profit, number_of_trades, average_loss, average_win

In [15]:
loss_win_trades_profit(USDJPY)

(-2471.0000000000127, 587, -78.57114754098359, 76.76142857142851)

#### Hypothesis testing
Question: How possible is it that the results of my trading strategy are out of coincidence?
It is clear that it is impossible to answer this question with 100% certainty. Therefore, one has to set up a significance level as a decision point to accept or reject the strategy.

H0: Random-Walk theory is true, market price is unpredictable and therefore expected mean value for profit is 0.
    expected profit value = 0
    
H1: Random-Walk theory is not true, with this trading strategy the mean profit is bigger or lower 0
    expected profit value <> 0

significance level: the resulting profit of this trading strategy has to exceed 3.9 standard deviations. This means that the occured mean profit of this strategy has to be 3.9 standard deviations different from the expected mean profit of 0 from the H0 to accept H1 and reject H0. In math: Sigma > 3.9

Reference for my considerations: https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2474755

In [16]:
def testStrategy(df):
    sum_of_profit = loss_win_trades_profit(df)[0]
    number_of_trades = loss_win_trades_profit(df)[1]
    average_loss = loss_win_trades_profit(df)[2]
    average_win = loss_win_trades_profit(df)[3]
    percentage_avg_win = abs(average_loss)/(abs(average_loss)+abs(average_win))
    percentage_avg_loss = 1 - percentage_avg_win
    exp_profit = 0
    variance_of_trades = (percentage_avg_loss*(average_loss-exp_profit)**2) + (percentage_avg_win*(average_win-exp_profit)**2)
    stand_dev_of_profit = math.sqrt(number_of_trades)*math.sqrt(variance_of_trades)
    stand_dev_final = sum_of_profit/stand_dev_of_profit
    
    return stand_dev_final

In [17]:
std_USDJPY = testStrategy(USDJPY)
std_USDJPY

-1.3132597155786911

#### Conclusion: 
The negative value for the standard deviation can be turned into +1.31, if the actions for going short and long would be switched. The resulting standard deviation is below the signifincance level of 3.9 sigma and H0 cannot be rejected.

In [18]:
#Store created USDJPY_daily with profit column as csv
#USDJPY.to_csv('./cleaned_datafiles/usdjpy_rev_candle.csv', index=False)

### Testing the strategy with minutes data for USDJPY to check if the performance is better

In [19]:
USDJPY_min = pd.read_sql_query('SELECT * FROM trading.usdjpy_minute', engine)
USDJPY_min

date     open     high      low    close
0        2007-01-02 00:31:00  23752.0  23753.0  23743.0  23746.0
1        2007-01-02 00:32:00  23746.0  23746.0  23742.0  23744.0
2        2007-01-02 00:33:00  23744.0  23744.0  23740.0  23742.0
3        2007-01-02 00:34:00  23742.0  23742.0  23722.0  23734.0
4        2007-01-02 00:35:00  23734.0  23738.0  23734.0  23734.0
...                      ...      ...      ...      ...      ...
5999380  2023-03-03 15:00:00  27208.0  27209.0  27197.0  27201.0
5999381  2023-03-03 15:01:00  27201.0  27252.0  27201.0  27223.0
5999382  2023-03-03 15:02:00  27223.0  27244.0  27223.0  27242.0
5999383  2023-03-03 15:03:00  27242.0  27266.0  27242.0  27252.0
5999384  2023-03-03 15:04:00  27252.0  27258.0  27247.0  27251.0

[5999385 rows x 5 columns]

In [20]:
USDJPY_min = calcOpenCloseDiff(USDJPY_min)
USDJPY_min

date     open     high      low    close  \
0        2007-01-02 00:31:00  23752.0  23753.0  23743.0  23746.0   
1        2007-01-02 00:32:00  23746.0  23746.0  23742.0  23744.0   
2        2007-01-02 00:33:00  23744.0  23744.0  23740.0  23742.0   
3        2007-01-02 00:34:00  23742.0  23742.0  23722.0  23734.0   
4        2007-01-02 00:35:00  23734.0  23738.0  23734.0  23734.0   
...                      ...      ...      ...      ...      ...   
5999380  2023-03-03 15:00:00  27208.0  27209.0  27197.0  27201.0   
5999381  2023-03-03 15:01:00  27201.0  27252.0  27201.0  27223.0   
5999382  2023-03-03 15:02:00  27223.0  27244.0  27223.0  27242.0   
5999383  2023-03-03 15:03:00  27242.0  27266.0  27242.0  27252.0   
5999384  2023-03-03 15:04:00  27252.0  27258.0  27247.0  27251.0   

         diff_open_close  
0                   -6.0  
1                   -2.0  
2                   -2.0  
3                   -8.0  
4                    0.0  
...                  ...  
5999380             -7.0  
5999381             22.0  
5999382             19.0  
5999383             10.0  
5999384             -1.0  

[5999385 rows x 6 columns]

#### Wanted to improve the classLongShort function in performance by using np.where and shift, but accidentally wrote a function what just compared to the fifth previous value and not every value of the last 5 values. Named the function classLongShortFifthValue and keep it for the case that I also want to test this as a strategy

In [21]:
def classLongShortFifthValue(df):
    df['low_shifted'] = df['low'].shift(periods=5)
    df['high_shifted'] = df['high'].shift(periods=5)

    cond1 = df['low'] < df['low_shifted']
    cond2 = df['high'] > df['high_shifted']
    cond3 = df['diff_open_close'] > 0
    cond4 = df['diff_open_close'] < 0

    list_class = np.where(cond1 & cond3, -1, np.where(cond2 & cond4, 1, 0))

    df['classification'] = list_class
    
    df_new = df.drop(['low_shifted', 'high_shifted'], axis=1)
    
    return df_new

In [22]:
%%time
classLongShortFifthValue(USDJPY_min)

Wall time: 425 ms


date     open     high      low    close  \
0        2007-01-02 00:31:00  23752.0  23753.0  23743.0  23746.0   
1        2007-01-02 00:32:00  23746.0  23746.0  23742.0  23744.0   
2        2007-01-02 00:33:00  23744.0  23744.0  23740.0  23742.0   
3        2007-01-02 00:34:00  23742.0  23742.0  23722.0  23734.0   
4        2007-01-02 00:35:00  23734.0  23738.0  23734.0  23734.0   
...                      ...      ...      ...      ...      ...   
5999380  2023-03-03 15:00:00  27208.0  27209.0  27197.0  27201.0   
5999381  2023-03-03 15:01:00  27201.0  27252.0  27201.0  27223.0   
5999382  2023-03-03 15:02:00  27223.0  27244.0  27223.0  27242.0   
5999383  2023-03-03 15:03:00  27242.0  27266.0  27242.0  27252.0   
5999384  2023-03-03 15:04:00  27252.0  27258.0  27247.0  27251.0   

         diff_open_close  classification  
0                   -6.0               0  
1                   -2.0               0  
2                   -2.0               0  
3                   -8.0               0  
4                    0.0               0  
...                  ...             ...  
5999380             -7.0               0  
5999381             22.0              -1  
5999382             19.0               0  
5999383             10.0               0  
5999384             -1.0               1  

[5999385 rows x 7 columns]

In [23]:
createOrderType(USDJPY_min)
createProfit(USDJPY_min)
loss_win_trades_profit(USDJPY_min)

(-23480.0, 1679402, -3.0176687347473985, 3.0539297970755)

In [24]:
std_USDJPY_min = testStrategy(USDJPY_min)
std_USDJPY_min

-5.96836607632827

### Conclusion:
Tested here accidentally another strategy, instead of comparing the current value for high and low with the 5 previous values, the function just compared with the fifth previous value. The result is a standard deviation of -5.96. If one would just switch the actions of when going long and short, the standard deviation value will get +5.96. This would be above the significance level of 3.9 and due to our rules acceptable as a valid trading strategy if one does not take the costs into account. If we take the costs per trade into acount (here asssuming 1 tick spread and 1 tick fees) then in reality we would not make any profit.

In [25]:
#Store created USDJPY_min with profit column as csv
#USDJPY_min.to_csv('./cleaned_datafiles/after_testing/usdjpy_min_rev_candle.csv', index=False)

#### Creation of the correct performance improved class function comparing the current high and low values with every value of the 5 previous values

In [26]:
def classLongShortImproved(df):
    df['low_shifted_5'] = df['low'].shift(periods=5)
    df['low_shifted_4'] = df['low'].shift(periods=4)
    df['low_shifted_3'] = df['low'].shift(periods=3)
    df['low_shifted_2'] = df['low'].shift(periods=2)
    df['low_shifted_1'] = df['low'].shift(periods=1)
    df['high_shifted_5'] = df['high'].shift(periods=5)
    df['high_shifted_4'] = df['high'].shift(periods=4)
    df['high_shifted_3'] = df['high'].shift(periods=3)
    df['high_shifted_2'] = df['high'].shift(periods=2)
    df['high_shifted_1'] = df['high'].shift(periods=1)

    cond1 = df['low'] < df['low_shifted_5']
    cond2 = df['low'] < df['low_shifted_4']
    cond3 = df['low'] < df['low_shifted_3']
    cond4 = df['low'] < df['low_shifted_2']
    cond5 = df['low'] < df['low_shifted_1']
    
    cond6 = df['high'] > df['high_shifted_5']
    cond7 = df['high'] > df['high_shifted_4']
    cond8 = df['high'] > df['high_shifted_3']
    cond9 = df['high'] > df['high_shifted_2']
    cond10 = df['high'] > df['high_shifted_1']
    
    cond11 = df['diff_open_close'] > 0
    cond12 = df['diff_open_close'] < 0

    list_class = np.where(cond1 & cond2 & cond3 & cond4 & cond5 & cond11, -1, 
                np.where(cond6 & cond7 & cond8 & cond9 & cond10 & cond12, 1, 0))

    df['classification'] = list_class
    
    df_new = df.drop(['low_shifted_5', 'low_shifted_4', 'low_shifted_3', 'low_shifted_2', 'low_shifted_1', 'high_shifted_5', 
                      'high_shifted_4', 'high_shifted_3', 'high_shifted_2', 'high_shifted_1'], axis=1)
    
    return df_new

### Testing USDJPY daily again to see if performance improved class function works correctly and gives the same results

In [27]:
USDJPY

date     open     high      low    close  diff_open_close  \
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0   
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0   
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0   
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0   
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0   
...         ...      ...      ...      ...      ...              ...   
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0   
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0   
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0   
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0   
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0   

      classification order_type  profit  
0                  0       none     0.0  
1                  0       none     0.0  
2                  0       none     0.0  
3                  0       none     0.0  
4                  0       none     0.0  
...              ...        ...     ...  
5076               1       none     0.0  
5077               0      short    -1.0  
5078               0       none     0.0  
5079               0       none     0.0  
5080               0       none     0.0  

[5081 rows x 9 columns]

In [28]:
USDJPY_test = USDJPY.drop(['classification', 'order_type', 'profit'], axis=1)
USDJPY_test

date     open     high      low    close  diff_open_close
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0
...         ...      ...      ...      ...      ...              ...
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0

[5081 rows x 6 columns]

In [29]:
USDJPY_test = classLongShortImproved(USDJPY_test)
USDJPY_test

date     open     high      low    close  diff_open_close  \
0    2005-01-03  20548.0  20690.0  20464.0  20564.0             16.0   
1    2005-01-04  20564.0  20958.0  20486.0  20888.0            324.0   
2    2005-01-05  20888.0  21004.0  20752.0  20820.0            -68.0   
3    2005-01-06  20820.0  21036.0  20776.0  20984.0            164.0   
4    2005-01-07  20984.0  21024.0  20772.0  20964.0            -20.0   
...         ...      ...      ...      ...      ...              ...   
5076 2023-02-27  27285.0  27312.0  27182.0  27231.0            -54.0   
5077 2023-02-28  27224.0  27383.0  27146.0  27225.0              1.0   
5078 2023-03-01  27211.0  27293.0  27051.0  27230.0             19.0   
5079 2023-03-02  27228.0  27420.0  27204.0  27350.0            122.0   
5080 2023-03-03  27345.0  27353.0  27160.0  27212.0           -133.0   

      classification  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
5076               1  
5077               0  
5078               0  
5079               0  
5080               0  

[5081 rows x 7 columns]

In [30]:
USDJPY_test = createOrderType(USDJPY_test)
USDJPY_test = createProfit(USDJPY_test)
loss_win_trades_profit(USDJPY_test)

(-2471.0000000000127, 587, -78.57114754098359, 76.76142857142851)

In [31]:
std_USDJPY_test = testStrategy(USDJPY_test)
std_USDJPY_test

-1.3132597155786911

### Conclusion:
Performance improved class function gives the same results for USDJPY daily data.

### Testing USDJPY minute again with correct performance improved class function

In [32]:
USDJPY_min_test = USDJPY_min.drop(['classification', 'order_type', 'profit'], axis=1)

In [33]:
USDJPY_min_test = USDJPY_min_test.drop(['low_shifted', 'high_shifted'], axis=1)
USDJPY_min_test

date     open     high      low    close  \
0        2007-01-02 00:31:00  23752.0  23753.0  23743.0  23746.0   
1        2007-01-02 00:32:00  23746.0  23746.0  23742.0  23744.0   
2        2007-01-02 00:33:00  23744.0  23744.0  23740.0  23742.0   
3        2007-01-02 00:34:00  23742.0  23742.0  23722.0  23734.0   
4        2007-01-02 00:35:00  23734.0  23738.0  23734.0  23734.0   
...                      ...      ...      ...      ...      ...   
5999380  2023-03-03 15:00:00  27208.0  27209.0  27197.0  27201.0   
5999381  2023-03-03 15:01:00  27201.0  27252.0  27201.0  27223.0   
5999382  2023-03-03 15:02:00  27223.0  27244.0  27223.0  27242.0   
5999383  2023-03-03 15:03:00  27242.0  27266.0  27242.0  27252.0   
5999384  2023-03-03 15:04:00  27252.0  27258.0  27247.0  27251.0   

         diff_open_close  
0                   -6.0  
1                   -2.0  
2                   -2.0  
3                   -8.0  
4                    0.0  
...                  ...  
5999380             -7.0  
5999381             22.0  
5999382             19.0  
5999383             10.0  
5999384             -1.0  

[5999385 rows x 6 columns]

In [34]:
USDJPY_min_test = classLongShortImproved(USDJPY_min_test)
USDJPY_min_test

date     open     high      low    close  \
0        2007-01-02 00:31:00  23752.0  23753.0  23743.0  23746.0   
1        2007-01-02 00:32:00  23746.0  23746.0  23742.0  23744.0   
2        2007-01-02 00:33:00  23744.0  23744.0  23740.0  23742.0   
3        2007-01-02 00:34:00  23742.0  23742.0  23722.0  23734.0   
4        2007-01-02 00:35:00  23734.0  23738.0  23734.0  23734.0   
...                      ...      ...      ...      ...      ...   
5999380  2023-03-03 15:00:00  27208.0  27209.0  27197.0  27201.0   
5999381  2023-03-03 15:01:00  27201.0  27252.0  27201.0  27223.0   
5999382  2023-03-03 15:02:00  27223.0  27244.0  27223.0  27242.0   
5999383  2023-03-03 15:03:00  27242.0  27266.0  27242.0  27252.0   
5999384  2023-03-03 15:04:00  27252.0  27258.0  27247.0  27251.0   

         diff_open_close  classification  
0                   -6.0               0  
1                   -2.0               0  
2                   -2.0               0  
3                   -8.0               0  
4                    0.0               0  
...                  ...             ...  
5999380             -7.0               0  
5999381             22.0               0  
5999382             19.0               0  
5999383             10.0               0  
5999384             -1.0               0  

[5999385 rows x 7 columns]

In [35]:
USDJPY_min_test = createOrderType(USDJPY_min_test)
USDJPY_min_test = createProfit(USDJPY_min_test)
loss_win_trades_profit(USDJPY_min_test)

(1054.0, 251714, -3.298647437223552, 3.3560431216217568)

In [36]:
std_USDJPY_min_test = testStrategy(USDJPY_min_test)
std_USDJPY_min_test

0.6314009180530585

### Conclusion:
The resulting standard deviation is below and the significance level of 3.9 sigma and therefore H0 cannot be rejected.

### Testing EURUSD daily

In [37]:
EURUSD = pd.read_sql_query('SELECT * FROM trading.eurusd', engine)
EURUSD

date     open     high      low    close
0    2005-03-09  26674.0  26910.0  26666.0  26864.0
1    2005-03-10  26864.0  26906.0  26794.0  26848.0
2    2005-03-11  26848.0  26956.0  26644.0  26914.0
3    2005-03-13  26948.0  26952.0  26886.0  26904.0
4    2005-03-14  26904.0  26910.0  26670.0  26724.0
...         ...      ...      ...      ...      ...
4630 2022-11-16  20673.0  20877.0  20661.0  20788.0
4631 2022-11-17  20780.0  20813.0  20610.0  20718.0
4632 2022-11-18  20715.0  20792.0  20627.0  20648.0
4633 2022-11-21  20643.0  20666.0  20445.0  20478.0
4634 2022-11-22  20475.0  20580.0  20475.0  20570.0

[4635 rows x 5 columns]

In [38]:
EURUSD = calcOpenCloseDiff(EURUSD)
EURUSD = classLongShortImproved(EURUSD)
EURUSD = createOrderType(EURUSD)
EURUSD = createProfit(EURUSD)
loss_win_trades_profit(EURUSD)

(-1313.0, 642, -107.30407523510972, 103.5125786163522)

In [39]:
std_EURUSD = testStrategy(EURUSD)
std_EURUSD

-0.49169131848834957

### Conclusion:
By switching the action for going long and short, the resulting standard deviation of +0.49 is below the significance level of 3.9 sigma and therefore H0 cannot be rejected.

### Testing EURUSD daily fifth value strategy

In [40]:
EURUSD_fifth = EURUSD.drop(['classification', 'order_type', 'profit'], axis=1)
EURUSD_fifth

date     open     high      low    close  diff_open_close
0    2005-03-09  26674.0  26910.0  26666.0  26864.0            190.0
1    2005-03-10  26864.0  26906.0  26794.0  26848.0            -16.0
2    2005-03-11  26848.0  26956.0  26644.0  26914.0             66.0
3    2005-03-13  26948.0  26952.0  26886.0  26904.0            -44.0
4    2005-03-14  26904.0  26910.0  26670.0  26724.0           -180.0
...         ...      ...      ...      ...      ...              ...
4630 2022-11-16  20673.0  20877.0  20661.0  20788.0            115.0
4631 2022-11-17  20780.0  20813.0  20610.0  20718.0            -62.0
4632 2022-11-18  20715.0  20792.0  20627.0  20648.0            -67.0
4633 2022-11-21  20643.0  20666.0  20445.0  20478.0           -165.0
4634 2022-11-22  20475.0  20580.0  20475.0  20570.0             95.0

[4635 rows x 6 columns]

In [41]:
EURUSD_fifth = classLongShortFifthValue(EURUSD_fifth)
EURUSD_fifth = createOrderType(EURUSD_fifth)
EURUSD_fifth = createProfit(EURUSD_fifth)
loss_win_trades_profit(EURUSD_fifth)

(-7362.0, 2019, -104.88867376573089, 103.79033915724563)

In [42]:
std_EURUSD_fifth = testStrategy(EURUSD_fifth)
std_EURUSD_fifth

-1.5703081728087107

### Conclusion:
The resulting standard deviation for switched action would still be under the significance level of 3.9 sigma and therefore H0 cannot be rejected

### Testing GBPUSD

In [43]:
GBPUSD = pd.read_sql_query('SELECT * FROM trading.gbpusd', engine)
GBPUSD

date     open     high      low    close
0    2005-03-09  38412.0  38596.0  38398.0  38548.0
1    2005-03-10  38550.0  38610.0  38408.0  38472.0
2    2005-03-11  38472.0  38562.0  38326.0  38514.0
3    2005-03-13  38534.0  38540.0  38440.0  38470.0
4    2005-03-14  38472.0  38494.0  38198.0  38310.0
...         ...      ...      ...      ...      ...
4589 2022-09-20  22837.0  22921.0  22714.0  22761.0
4590 2022-09-21  22727.0  22769.0  22471.0  22540.0
4591 2022-09-22  22428.0  22727.0  22423.0  22510.0
4592 2022-09-23  22493.0  22547.0  21678.0  21680.0
4593 2022-09-26  21615.0  21691.0  20614.0  21566.0

[4594 rows x 5 columns]

In [44]:
GBPUSD = calcOpenCloseDiff(GBPUSD)
GBPUSD = classLongShortImproved(GBPUSD)
GBPUSD = createOrderType(GBPUSD)
GBPUSD = createProfit(GBPUSD)
loss_win_trades_profit(GBPUSD)

(-7080.0, 567, -141.91304347826087, 133.40377358490565)

In [45]:
std_GBPUSD = testStrategy(GBPUSD)
std_GBPUSD

-2.1609589344956044

### Conclusion:
By switching the actions for long and short, the resulting standard deviation of +2.16 is below the significance level of 3.9 sigma and therefore H0 cannot be rejected.

### Testing GBPUSD daily fifth value strategy

In [49]:
GBPUSD_fifth = GBPUSD.drop(['classification', 'order_type', 'profit'], axis=1)
GBPUSD_fifth

date     open     high      low    close  diff_open_close
0    2005-03-09  38412.0  38596.0  38398.0  38548.0            136.0
1    2005-03-10  38550.0  38610.0  38408.0  38472.0            -78.0
2    2005-03-11  38472.0  38562.0  38326.0  38514.0             42.0
3    2005-03-13  38534.0  38540.0  38440.0  38470.0            -64.0
4    2005-03-14  38472.0  38494.0  38198.0  38310.0           -162.0
...         ...      ...      ...      ...      ...              ...
4589 2022-09-20  22837.0  22921.0  22714.0  22761.0            -76.0
4590 2022-09-21  22727.0  22769.0  22471.0  22540.0           -187.0
4591 2022-09-22  22428.0  22727.0  22423.0  22510.0             82.0
4592 2022-09-23  22493.0  22547.0  21678.0  21680.0           -813.0
4593 2022-09-26  21615.0  21691.0  20614.0  21566.0            -49.0

[4594 rows x 6 columns]

In [50]:
GBPUSD_fifth = classLongShortFifthValue(GBPUSD_fifth)
GBPUSD_fifth = createOrderType(GBPUSD_fifth)
GBPUSD_fifth = createProfit(GBPUSD_fifth)
loss_win_trades_profit(GBPUSD_fifth)

(2551.0, 1964, -128.3829568788501, 130.06727828746176)

In [51]:
std_GBPUSD_fifth = testStrategy(GBPUSD_fifth)
std_GBPUSD_fifth

0.4454531184435549

### Conclusion:
The resulting standard deviation is below the significance level of 3.9 sigma and therefore H0 cannot be rejected

### Testing USDCAD

In [52]:
USDCAD = pd.read_sql_query('SELECT * FROM trading.usdcad', engine)
USDCAD

date     open     high      low    close
0    2005-01-03  24070.0  24246.0  24010.0  24154.0
1    2005-01-04  24154.0  24538.0  24114.0  24450.0
2    2005-01-05  24450.0  24656.0  24418.0  24514.0
3    2005-01-06  24514.0  24788.0  24472.0  24708.0
4    2005-01-07  24708.0  24762.0  24458.0  24634.0
...         ...      ...      ...      ...      ...
5067 2023-02-27  27149.0  27248.0  27067.0  27149.0
5068 2023-02-28  27112.0  27296.0  27112.0  27293.0
5069 2023-03-01  27276.0  27318.0  27168.0  27180.0
5070 2023-03-02  27177.0  27282.0  27164.0  27190.0
5071 2023-03-03  27169.0  27289.0  27110.0  27243.0

[5072 rows x 5 columns]

In [53]:
USDCAD = calcOpenCloseDiff(USDCAD)
USDCAD = classLongShortImproved(USDCAD)
USDCAD = createOrderType(USDCAD)
USDCAD = createProfit(USDCAD)
loss_win_trades_profit(USDCAD)

(-650.2000000000153, 654, -94.20813953488374, 102.4432258064516)

In [54]:
std_USDCAD = testStrategy(USDCAD)
std_USDCAD

-0.25880475700003047

### Conclusion: 
The standard deviation is below the significance level of 3.9 and therefore H0 cannot be rejected

### Testing USDCAD daily fifth value strategy

In [55]:
USDCAD_fifth = USDCAD.drop(['classification', 'order_type', 'profit'], axis=1)
USDCAD_fifth

date     open     high      low    close  diff_open_close
0    2005-01-03  24070.0  24246.0  24010.0  24154.0             84.0
1    2005-01-04  24154.0  24538.0  24114.0  24450.0            296.0
2    2005-01-05  24450.0  24656.0  24418.0  24514.0             64.0
3    2005-01-06  24514.0  24788.0  24472.0  24708.0            194.0
4    2005-01-07  24708.0  24762.0  24458.0  24634.0            -74.0
...         ...      ...      ...      ...      ...              ...
5067 2023-02-27  27149.0  27248.0  27067.0  27149.0              0.0
5068 2023-02-28  27112.0  27296.0  27112.0  27293.0            181.0
5069 2023-03-01  27276.0  27318.0  27168.0  27180.0            -96.0
5070 2023-03-02  27177.0  27282.0  27164.0  27190.0             13.0
5071 2023-03-03  27169.0  27289.0  27110.0  27243.0             74.0

[5072 rows x 6 columns]

In [56]:
USDCAD_fifth = classLongShortFifthValue(USDCAD_fifth)
USDCAD_fifth = createOrderType(USDCAD_fifth)
USDCAD_fifth = createProfit(USDCAD_fifth)
loss_win_trades_profit(USDCAD_fifth)

(-3900.8000000000284, 2192, -88.53473315835517, 93.73256262042389)

In [57]:
std_USDCAD_fifth = testStrategy(USDCAD_fifth)
std_USDCAD_fifth

-0.9146002639085394

### Conclusion:
The standard deviation is below the significance level of 3.9 and therefore H0 cannot be rejected